lib/utmp.c

```diff
@@ -161,16 +161,23 @@ get_current_utmp(pid_t main_pid)
 		    && (USER_PROCESS  != ut->ut_type))
 			continue;
 
-		if (   (main_pid == ut->ut_pid)
-		    && ('\0' != ut->ut_id[0])
-		    /* A process may have failed to close an entry
-		     * Check if this entry refers to the current tty */
-		    && is_my_tty(ut->ut_line))
-		{
-			break;
+		if (main_pid == ut->ut_pid) {
+			if (is_my_tty(ut->ut_line))
+				break; /* Perfect match, stop the search */
+
+			if (NULL == ut_by_pid)
+				ut_by_pid = ut;
+
+		} else if (   (NULL == ut_by_line)
+			   && (LOGIN_PROCESS == ut->ut_type) /* Be more picky when matching by 'ut_line' only */
+			   && (is_my_tty(ut->ut_line))) {
+			ut_by_line = ut;
 		}
 	}
 
+	if (NULL == ut)
+		ut = ut_by_pid ?: ut_by_line;
```

Comment:
We can avoid the conditional with a one-liner:

```c
ut = ut ?: ut_by_pid ?: ut_by_line;
```

Reply:
You asked earlier to expand the loop to brea/continue version, which has more lines, but better readability.
The same is here.
Unless we have next level of folding, like 
```C
ut ?:= ut_by_pid ?: ut_by_line;
```
my version has better readability.

Reply:
> ```c
> ut ?:= ut_by_pid ?: ut_by_line;
> ```

You may discuss this operation on the C committee.